duckdb/duckdb
# implement SERIAL

Fixes https://github.com/duckdb/duckdb/issues/1768

Based on [#3599](https://github.com/duckdb/duckdb/pull/3599) and [#4338](https://github.com/duckdb/duckdb/pull/4338) (thanks @jwills and @ila),

```sql
CREATE TABLE tablefoo (
    bar serial
);
```
Basically, do the following things:

```sql 
CREATE SEQUENCE tablefoo_bar_seq;
CREATE TABLE tablefoo (
    bar integer NOT NULL DEFAULT nextval('tablefoo_bar_seq')
);
ALTER SEQUENCE tablefoo_bar_seq OWNED BY tablefoo;
```

This is a minimal working implementation (passing all tests from [#3599](https://github.com/duckdb/duckdb/pull/3599)).

However, I need some help/guidance to complete the implementation !  🤝
Maybe @Tishj  @lnkuiper  can give me some advice.

1.  If I enable enable_verification in `test/sql/types/serial/test_serial_table.test`, something breaks (see FIXME in the test). 
How should I debug the enable_verification pragma?

2.  If I restart the database in `test/sql/storage/catalog/test_serial_storage.test`, the sequence does not drop with the DROP TABLE statement (see FIXME in the test). How should I debug the storage code (I think there may be a missing dependency where the catalog doesn't persist in the db) ?

for example 
```sql
✅
CREATE TABLE tablefoo (
    bar serial
);

DROP TABLE tablefoo
------------------------------Restart DB  ------------------------------
CREATE TABLE tablefoo (
    bar serial
);


❌
CREATE TABLE tablefoo (
    bar serial
);
------------------------------Restart DB  ------------------------------
DROP TABLE tablefoo

CREATE TABLE tablefoo (
    bar serial
);
```
3. Is my implementation on the right track?

TODO:
1. Fix the broken test.
2. Implement all Postgres compatible ALTER statements (e.g., rename, change type) for the serial column.
